examples/application/main.go

```diff
@@ -0,0 +1,156 @@
+	if err != nil {
+		log.Printf("Error removing Application: %s", err)
+		err = nil
+	}
+	err = client.AppPools.Delete(appPoolName)
+	if err != nil {
+		log.Printf("Error removing App Pool: %s", err)
+		err = nil
+	}
+	err = os.RemoveAll(testAppPath)
+	if err != nil {
+		log.Printf("Error removing test directory: %s", err)
+		err = nil
+	}
+}
+
+func testApplication(testUrl string, testAppContent string) {
+	resp, err := http.Get(testUrl)
+	if err != nil {
+		log.Printf("Error trying to get the test HTML from IIS: %s", err)
+		err = nil
+	} else {
+		respContent, err := ioutil.ReadAll(resp.Body)
+		if err != nil {
+			log.Printf("Error getting HTTP response body: %s", err)
+			err = nil
+		} else {
+			log.Printf("Got test HTML content: %q", respContent)
+
+			diff := strings.Compare(testAppContent, string(respContent))
+			log.Printf("Difference between what was written and read: %d", diff)
+		}
+	}
+}
+
+func createApplication(client *iis.Client, applicationName string, appPoolName string, testAppPath string, websiteName string) error {
+	err := client.Applications.Create(applicationName, appPoolName, testAppPath, websiteName)
+	if err != nil {
+		_ = client.AppPools.Delete(appPoolName)
+		_ = os.RemoveAll(testAppPath)
+		return fmt.Errorf("Error creating IIS Application: %s", err)
+	}
+
+	exists, err := client.Applications.Exists(applicationName, websiteName)
+	if err != nil {
+		_ = client.AppPools.Delete(appPoolName)
+		_ = os.RemoveAll(testAppPath)
+		return fmt.Errorf("Error checking for IIS Application existence: %s", err)
+	}
+	if !*exists {
+		return fmt.Errorf("Tried to create Application %q under Website %q but it doesn't exist! %s", applicationName, websiteName, err)
+	}
+	log.Printf("Application created successfully!")
+
+	return nil
+}
+
+func createAppPool(client *iis.Client, appPoolName string) error {
+	err := client.AppPools.Create(appPoolName)
+	if err != nil {
+		return fmt.Errorf("Error creating IIS App Pool: %s", err)
+	}
+
+	exists, err := client.AppPools.Exists(appPoolName)
+	if err != nil {
+		return fmt.Errorf("Error checking for IIS App Pool existence: %s", err)
+	}
+	if !*exists {
+		return fmt.Errorf("Tried to create an app pool %q but it doesn't exist! %s", appPoolName, err)
+	}
+	log.Printf("App Pool created successfully!")
+
+	return nil
+}
+
+func createTestDirectory(testAppPath string, testAppContent string) error {
+	log.Printf("Creating test application folder %q", testAppPath)
+	err := os.MkdirAll(testAppPath, os.ModeDir)
+	if err != nil {
+		return fmt.Errorf("Error creating directory %q: %s", testAppPath, err)
+	}
+
+	log.Printf("Writing test HTML file")
+	testFile, err := os.OpenFile(fmt.Sprintf("%s\\index.html", testAppPath), os.O_CREATE|os.O_WRONLY, os.ModePerm)
+	_, err = testFile.WriteString(testAppContent)
+	if err != nil {
+		_ = testFile.Close()
+		_ = os.RemoveAll(testAppPath)
+		return fmt.Errorf("Error writing test HTML file: %s", err)
+	}
+	_ = testFile.Close()
```

Comment:
rather than doing this manually, we can use a `defer` here to ensure this is always run e.g.

```
func createTestDirectory(testAppPath string, testAppContent string) error {
	log.Printf("Creating test application folder %q", testAppPath)
	err := os.MkdirAll(testAppPath, os.ModeDir)
	if err != nil {
		return fmt.Errorf("Error creating directory %q: %s", testAppPath, err)
	}

	log.Printf("Writing test HTML file")
	testFile, err := os.OpenFile(fmt.Sprintf("%s\\index.html", testAppPath), os.O_CREATE|os.O_WRONLY, os.ModePerm)
	defer testFile.Close()
	_, err = testFile.WriteString(testAppContent)
	if err != nil {
		_ = os.RemoveAll(testAppPath)
		return fmt.Errorf("Error writing test HTML file: %s", err)
	}
	log.Printf("Wrote test HTML content: %q", testAppContent)

	return nil
}
```